packages/berry-pnpify/sources/NodeModulesFS.ts

```diff
@@ -0,0 +1,253 @@
+import { CreateReadStreamOptions, CreateWriteStreamOptions } from '@berry/fslib';
+import { NodeFS, FakeFS, WriteFileOptions }                  from '@berry/fslib';
+
+import fs                                                    from 'fs';
+
+import { NodePathResolver }                                  from './NodePathResolver';
+import { PnPApiLoader }                                      from './PnPApiLoader';
+import { PnPApiLocator }                                     from './PnPApiLocator';
+
+export type NodeModulesFSOptions = {
+  baseFs?: FakeFS
+};
+
+export class NodeModulesFS extends FakeFS {
+  private readonly baseFs: FakeFS;
+  private readonly pathResolver: NodePathResolver;
+
+  constructor({baseFs = new NodeFS()}: NodeModulesFSOptions = {}) {
+    super();
+
+    this.baseFs = baseFs;
+    this.pathResolver = new NodePathResolver({
+      apiLoader: new PnPApiLoader({ watch: (fs as any).watch.bind(fs) }),
+      apiLocator: new PnPApiLocator({ existsSync: baseFs.existsSync.bind(baseFs) })
+    });
+  }
+
+  getRealPath() {
+    return '/';
+  }
+
+  getBaseFs() {
+    return this.baseFs;
+  }
+
+  private resolveFilePath(p: string): string {
+    const pnpPath = this.pathResolver.resolvePath(p);
+    if (!pnpPath.resolvedPath) {
+      throw new Error(`ENOENT: no such file or directory, stat '${p}'`)
+    } else {
+      return pnpPath.resolvedPath;
+    }
+  }
+
+  private resolveStatPath(p: string): string {
```

Comment:
Why is `resolveStatPath` different from `resolveFilePath`? I saw the comment in `NodePathResolver` but it's still not super clear to me why this `statPath` needs to be different from `resolvedPath` 🤔

Reply:
@arcanis Dirs ending with '/node_modules/foo/node_modules' or '.../node_modules/foo/node_modules/@scope' do not physically exist, but we must pretend they do exist if package `foo` has dependencies and there is some package `@scope/bar` in these dependencies. We need two things to emulate their existence:
1. List of entries in these dirs. We can have them by calling PnP API and getting dependencies for the issuer `.../foo/`
2. And we need either fake stats or we can forward underlying fs to stat the issuer dir. The issuer dir exists on fs.